addons/calendar/models/calendar.py

```diff
@@ -0,0 +1,1047 @@
+
+            :param string date: date/time to be formatted
+            :param string interval: Among 'day', 'month', 'dayname' and 'time' indicating the desired formatting
+            :param string tz: Timezone indicator (optional)
+
+            :return unicode: Formatted date or time (as unicode string, to prevent jinja2 crash)
+
+            (Function used only in calendar_event_data.xml) '''
+
+        date = fields.Datetime.from_string(date)
+
+        if tz:
+            timezone = pytz.timezone(tz or 'UTC')
+            date = date.replace(tzinfo=pytz.timezone('UTC')).astimezone(timezone)
+
+        if interval == 'day':
+            # Day number (1-31)
+            res = unicode(date.day)
+
+        elif interval == 'month':
+            # Localized month name and year
+            res = babel.dates.format_date(date=date, format='MMMM y')
+
+        elif interval == 'dayname':
+            # Localized day name
+            res = babel.dates.format_date(date=date, format='EEEE')
+
+        elif interval == 'time':
+            # Localized time
+
+            dummy, format_time = self.get_date_formats()
+            res = tools.ustr(date.strftime(format_time + " %Z"))
+
+        return res
+
+    @api.model
+    def search(self, args, offset=0, limit=0, order=None, count=False):
+
+        if self.env.context.get('mymeetings'):
+            args += [('partner_ids', 'in', [self.env.user.partner_id.id])]
+
+        new_args = []
+        for arg in args:
+            new_arg = arg
+
+            if arg[0] in ('start_date', 'start_datetime', 'start',) and arg[1] == ">=":
+                if self.env.context.get('virtual_id', True):
+                    new_args += ['|', '&', ('recurrency', '=', 1), ('final_date', arg[1], arg[2])]
+            elif arg[0] == "id":
+                new_id = get_real_ids(arg[2])
+                new_arg = (arg[0], arg[1], new_id)
+            new_args.append(new_arg)
+
+        if not self.env.context.get('virtual_id', True):
+            return super(CalendarEvent, self).search(new_args, offset=offset, limit=limit, order=order, count=count)
+        # offset, limit, order and count must be treated separately as we may need to deal with virtual ids
+        res = super(CalendarEvent, self).search(new_args, offset=0, limit=0, order=None, count=False)
+        res = res.get_recurrent_ids(args, order=order)
+        if count:
+            return len(res)
+        elif limit:
+            return res[offset: offset + limit]
+        return res
+
+    @api.multi
+    def copy(self, default=None):
+        self.ensure_one()
+        default = dict(default or {})
+        self._set_date(default, id=default.get('id'))
+        return super(CalendarEvent, self).copy(default)
+
+    def _detach_one_event(self, values=dict()):
+        real_event_id = calendar_id2real_id(self.id)
+        data = self.read(['allday', 'start', 'stop', 'rrule', 'duration'])[0]
```

Comment:
If self is already a singleton then why not getting values of 'allday', 'start', 'stop', 'rrule', 'duration' instead of reading it ?